packages/widget/src/controllers/transfers/fungibleTokenTransfer.ts

```diff
@@ -0,0 +1,284 @@
+    }
+    if (this.destinatonAddress === '') {
+      return FungibleTransferState.MISSING_DESTINATION_ADDRESS;
+    }
+    if (
+      !this.walletContext.value?.evmWallet &&
+      !this.walletContext.value?.substrateWallet
+    ) {
+      return FungibleTransferState.WALLET_NOT_CONNECTED;
+    }
+    if (
+      this.sourceNetwork.type === Network.EVM &&
+      this.walletContext.value?.evmWallet?.providerChainId !==
+        this.sourceNetwork.chainId
+    ) {
+      return FungibleTransferState.WRONG_CHAIN;
+    }
+    if (this.waitingUserConfirmation) {
+      return FungibleTransferState.WAITING_USER_CONFIRMATION;
+    }
+    if (this.waitingTxExecution) {
+      return FungibleTransferState.WAITING_TX_EXECUTION;
+    }
+    if (this.transferTransactionId) {
+      return FungibleTransferState.COMPLETED;
+    }
+    if (this.pendingEvmApprovalTransactions.length > 0) {
+      return FungibleTransferState.PENDING_APPROVALS;
+    }
+    if (this.pendingEvmTransferTransaction) {
+      return FungibleTransferState.PENDING_TRANSFER;
+    }
+    return FungibleTransferState.UNKNOWN;
+  }
+
+  executeTransaction(): void {
+    if (!this.sourceNetwork) {
+      return;
+    }
+    switch (this.sourceNetwork.type) {
+      case Network.EVM:
+        {
+          void this.executeNextEvmTransaction();
+        }
+        break;
+      case Network.SUBSTRATE:
+        {
+          //TODO: add substrate logic
+        }
+        break;
+      default:
+        throw new Error('Unsupported network type');
+    }
+  }
+
+  getExplorerLink(): string {
+    if (this.env === Environment.MAINNET) {
+      return `https://scan.buildwithsygma.com/transfer/${this.transferTransactionId}`;
+    }
+    return `https://scan.test.buildwithsygma.com/transfer/${this.transferTransactionId}`;
+  }
+
+  private filterDestinationNetworksAndResources = async (
+    sourceNetwork: Domain
+  ): Promise<void> => {
+    if (!this.routesCache.has(sourceNetwork.chainId)) {
+      this.routesCache.set(
+        sourceNetwork.chainId,
+        await getRoutes(this.env, sourceNetwork.chainId, 'fungible')
+      );
+    }
+    if (!this.destinationNetwork) {
+      this.supportedDestinationNetworks = [];
+    }
+    this.supportedResources = [];
+    this.routesCache.get(sourceNetwork.chainId)!.forEach((route: Route) => {
+      //change/add dst routes if dst not yet selected, not a src domain and isn't already present
+      if (
+        !this.destinationNetwork &&
+        this.sourceNetwork?.chainId !== route.toDomain.chainId &&
+        !this.supportedDestinationNetworks.includes(route.toDomain)
+      ) {
+        this.supportedDestinationNetworks.push(route.toDomain);
+      }
+      if (
+        !this.destinationNetwork ||
+        this.destinationNetwork.chainId === route.toDomain.chainId
+      ) {
+        if (!this.supportedResources.includes(route.resource)) {
+          this.supportedResources.push(route.resource);
+        }
+      }
+    });
```

Comment:
i find it way cleaner to use filter/map than going through array with forEach and manually pushing to some other array

```suggestion
    this.supportedResources = [];

    if (!this.destinationNetwork) {
      this.supportedDestinationNetworks = this.routesCache
        .get(sourceNetwork.chainId)!
        .filter(
          (route) =>
            route.toDomain.chainId !== sourceNetwork.chainId &&
            !this.supportedDestinationNetworks.includes(route.toDomain)
        )
        .map((route) => route.toDomain);
    }

    this.supportedResources = this.routesCache
      .get(sourceNetwork.chainId)!
      .filter(
        (route) =>
          !this.destinationNetwork ||
          (route.toDomain.chainId === this.destinationNetwork?.chainId &&
            !this.supportedResources.includes(route.resource))
      )
      .map((route) => route.resource);
```

Reply:
ah, professional deformation! 😂 I tend to look at things in terms of performance (less repetitive loops the better) but it doesn't matter here.